Fix CPU Annotation Overlapping with Python Events

sraikund16 opened the pull request:
Summary:
Currently we have an issue where CPU User annotations can overlap with python events in the event that a python event calls step() within the function itself. To combat this, we can move the left side of the user annotation to the beginning of the parent python function. We do this because when instantiating the profiler we already start on step 0.
To implement this, we start by collecting all instances of ProfilerStep during post processing. Since TorchOps and Python events are sorted already, we can easily check if the current python event partially overlaps with the current ProfilerStep and, if so, alter the start time of the current ProfilerStep. We then move to the next ProfilerStep and continue iterating through all the python events. This keeps the time complexity of adding events to 'out' at O(s + n) -> O(n) post sorting, where "s" is the number of ProfilerSteps and "n" is the length of all events.

Test Plan:
Added unit test in which step() is called midway through a function. Afterwards, we print out a trace and then load the json to check that there are no overlaps. Also make sure that there is no regression in performance.

aaronenyeshi reviewed (changes requested):
try using LLONG_MAX, which is more portable, available in C and C++11

sraikund16 commented:
> try using LLONG_MAX, which is more portable, available in C and C++11

done, looks good now!

aaronenyeshi reviewed (approved):
lgtm, please make sure unit tests are passing

sraikund16 committed:
Fix CPU Annotation Overlapping with Python Events (#129599)

Summary:
Currently we have an issue where CPU User annotations can overlap with python events in the event that a python event calls step() within the function itself. To combat this, we can move the left side of the user annotation to the beginning of the parent python function. We do this because when instantiating the profiler we already start on step 0.

To implement this, we start by collecting all instances of ProfilerStep during post processing. Since TorchOps and Python events are sorted already, we can easily check if the current python event partially overlaps with the current ProfilerStep and, if so, alter the start time of the current ProfilerStep. We then move to the next ProfilerStep and continue iterating through all the python events. This keeps the time complexity of adding events to 'out' at O(s + n) -> O(n) post sorting, where "s" is the number of ProfilerSteps and "n" is the length of all events.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/129599

Test Plan:
Added unit test in which step() is called midway through a function. Afterwards, we print out a trace and then load the json to check that there are no overlaps

__Servicelab Metrics__

**__This Diff__**

**kineto_resnet**
    duration_ms: 14941
    number_of_events: 26752
    profiler_prepare_call_duration_us: 749
    profiler_enable_call_duration_us: 18482
    profiling_window_duration_us: 1828549
    profiler_disable_call_duration_us: 183867
    parse_kineto_call_duration_us: 909887
    function_events_build_tree_call_duration_us: 335641

**mlp_training**
    duration_ms: 59451
    number_of_events: 77856
    profiler_prepare_call_duration_us: 812
    profiler_enable_call_duration_us: 19854
    profiling_window_duration_us: 4779233
    profiler_disable_call_duration_us: 1771646
    parse_kineto_call_duration_us: 5353954
    function_events_build_tree_call_duration_us: 924587

 **mlp_training_cuda_graph**
    duration_ms: 12244
    number_of_events: 190
    profiler_prepare_call_duration_us: 781
    profiler_enable_call_duration_us: 18314
    profiling_window_duration_us: 206353
    profiler_disable_call_duration_us: 9821
    parse_kineto_call_duration_us: 6655
    function_events_build_tree_call_duration_us: 841

**__Main__**

**resnet_integration**
    duration_ms: 14898
    number_of_events: 26750
    profiler_prepare_call_duration_us: 647
    profiler_enable_call_duration_us: 338701
    profiling_window_duration_us: 1765469
    profiler_disable_call_duration_us: 167778
    parse_kineto_call_duration_us: 1018040
    function_events_build_tree_call_duration_us: 329391

 **mlp training**
    duration_ms: 57753
    number_of_events: 77821
    profiler_prepare_call_duration_us: 6140
    profiler_enable_call_duration_us: 19047
    profiling_window_duration_us: 5127060
    profiler_disable_call_duration_us: 1708149
    parse_kineto_call_duration_us: 5127904
    function_events_build_tree_call_duration_us: 879799

 **mlp_training_cuda_graph**
    duration_ms: 13095
    number_of_events: 184
    profiler_prepare_call_duration_us: 824
    profiler_enable_call_duration_us: 19980
    profiling_window_duration_us: 234567
    profiler_disable_call_duration_us: 11411
    parse_kineto_call_duration_us: 7810
    function_events_build_tree_call_duration_us: 1170

Reviewed By: aaronenyeshi

Differential Revision: D59076021
